jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java

```diff
@@ -65,6 +67,30 @@
         ASTERISK
     }
 
+    /**
+     * The concept of URI path parameters was originally specified in
+     * <a href="https://tools.ietf.org/html/rfc2396#section-3.3">RFC2396</a>, but that was
+     * obsoleted by
+     * <a href="https://tools.ietf.org/html/rfc3986#section-3.3">RFC3986</a> which removed
+     * a normative definition of path parameters. Specifically it excluded them from the
+     * <a href="https://tools.ietf.org/html/rfc3986#section-5.2.4">Remove Dot Segments</a>
+     * algorithm.  This results in some ambiguity as dot segments can result from later
+     * parameter removal or % encoding expansion, that are not removed from the URI
+     * by {@link URIUtil#canonicalPath(String)}.  Thus this class flags such ambiguous
+     * path segments, so that they may be rejected by the server if so configured.
+     */
+    private static final Trie<Boolean> __ambiguousSegments = new ArrayTrie<>();
+
+    static
+    {
+        __ambiguousSegments.put("%2e", Boolean.TRUE);
+        __ambiguousSegments.put("%2e%2e", Boolean.TRUE);
```

Comment:
This specific case `%2e%2e` I have especially difficult time accepting.
As that can show up as valid.
We even have testcases for it in DefaultServletTest. (see "testListingContextBreakout")
When the segment is 100% encoded I think it's fair and legit use.
However I do agree that when it's mixed (encoded / not-encoded) it's not valid and ambiguous.

Reply:
They are all fair usages.  They are just ambiguous because when we give a servlet a string version, we have thrown away information.
I think it is reasonable for a webapp that wants to consume `%2e%2e` to turn on the compliance mode to say they have their big boy pants on and are prepared to deal with that ambiguity.  But if an app doesn't do that, then it is best to not allow that request through.

Reply:
Too be more precise, I don't think we should be making a distinction between very ambiguous and slightly ambiguous URIs.   The compliance mode should simply exclude URIs that once we decode them and strip parameters it is impossible to recover exactly what the canonical segment boundaries are.

For me, the switching off that mode should be done by applications that will look at the raw getRequestURI and do their own parsing and decoding.  Note also with this change, we no longer strip intermediary parameters from the raw getRequestURI, so apps that want those can now get them.